### PR TITLE
Enable mask for AdamW optimizer

### PR DESCRIPTION
**Pull request by RissyRan**
# Description

Enable mask for AdamW optimizer
* Add mask option for AdamW optimizer
* Rename muon_test as optimizers_test, and add unit tests for mask generation

# Tests

* Add unit tests accordingly
* Expect all github runners are green

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/optimizers/optimizers.py [hunk 1]
```diff
@@ -23,6 +24,24 @@
...
+    def _is_decayed(path, _):
+      # Join path keys into a single string for pattern matching (e.g., "layer1/bias")
+      path_str = "/".join(str(p.key if hasattr(p, "key") else p.idx if hasattr(p, "idx") else p) for p in path)
+      # If any pattern in adamw_mask matches the path, exclude from weight decay (return False).
+      # Otherwise, apply weight decay (return True).
+      return not any(re.search(pattern, path_str) for pattern in config.adamw_mask)
+
+    return jax.tree_util.tree_map_with_path(_is_decayed, params)
+
+  return mask_fn
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply src/maxtext/optimizers/optimizers.py:44]: Updated.

File: tests/unit/optimizers_test.py [hunk 2]
```diff
@@ -236,5 +243,109 @@ def test_model_integration(self, model_name, expected_output):
...
+    argv = ["", get_test_config_path(), "run_name=test", "adamw_mask=['[']"]
+    config = pyconfig.initialize(argv)
+    mask_fn = optimizers.get_adamw_mask(config)
+    self.assertTrue(callable(mask_fn))
+
+    params = {"layer1": {"kernel": 1}}
+
+    # Applying the mask should raise re.error due to the invalid regex
+    with self.assertRaises(re.error):
+      mask_fn(params)
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply tests/unit/optimizers_test.py]: Updated.

File: src/maxtext/optimizers/optimizers.py [hunk 3]
```diff
@@ -23,6 +24,26 @@
...
+  """Create a mask function for AdamW optimizer to exclude certain parameters from weight decay."""
+  if not getattr(config, "adamw_mask", None):
+    return None
+
+  compiled_patterns = [re.compile(pattern) for pattern in config.adamw_mask]
+
+  def mask_fn(params):
+    def _is_decayed(path, _):
+      # Join path keys into a single string for pattern matching (e.g., "layer1/bias")
+      path_str = "/".join(str(p.key if hasattr(p, "key") else p.idx if hasattr(p, "idx") else p) for p in path)
```
Comment (gagika) [inline src/maxtext/optimizers/optimizers.py]: Gemini also suggested adding "name" attribute: 

JAX paths can also include GetAttrKey (e.g., when traversing custom PyTree objects), which uses a .name attribute. If it falls back to str(p), it might produce an ugly string representation like <jax.tree_util.GetAttrKey object at ...>.

To be completely bulletproof, consider adding .name to the fallback chain:

Python
path_str = "/".join(
    str(getattr(p, "key", getattr(p, "idx", getattr(p, "name", p)))) 
    for p in path
)

File: src/maxtext/configs/base.yml [hunk 4]
```diff
@@ -779,6 +779,7 @@ adam_b2: 0.95 # Exponential decay rate to track the second moment of past gradie
 adam_eps: 1.e-8 # A small constant applied to denominator outside of the square root.
 adam_eps_root: 0. # A small constant applied to denominator inside the square root.
 adam_weight_decay: 0.1 # AdamW Weight decay
+adamw_mask: [] # List of parameter names/patterns to exclude from weight decay in AdamW
```
Comment (gagika) [inline src/maxtext/configs/base.yml]: nit: could you provide an example here: e.g. ['.*norm', 'bias'] ?

Comment (RissyRan) [reply src/maxtext/configs/base.yml]: Sounds good, and will add.

File: src/maxtext/optimizers/optimizers.py (hunk 3 above)
Comment (RissyRan) [reply src/maxtext/optimizers/optimizers.py]: Sounds good, updated accordingly.
